### ISSUE TITLE
Improve handling of Apple filesystem layouts

### ISSUE THREAD

**Issue by bitsgalore** [KBNLresearch/isolyzer#22]
Image here:

https://archive.org/download/691-6551-A2ZiWork_v9.0.3._Install_Disc_2009_DVD

For the .toast image here isolyzer doesn't recognize any file system,  but a look in a Hex editor shows a  0x4552 byte pattern at the start of the file. This suggests this is an Apple Partition map (see also [here](https://github.com/KBNLresearch/isolyzer/blob/07f74991bd22b7f21b5f7e6af3625dfb4cb28ced/isolyzer/isolyzer.py#L3040)). According to Isolyzer's current parsing logic this pattern should be accompanied by a 0x504D byte sequence at offset 512, which is not there in this case. But I DO see this pattern at offset 2048 (i.e. the start of the 2nd sector), and there's an "Apple_HFS" file system further on. So I suspect this is some specific Apple layout that is not currently suported/handled.

**Comment by thorsted** [comment]
Looks like offset 2048 is for partitions with a driver map, probably for making disc bootable.
https://opensource.apple.com/source/IOStorageFamily/IOStorageFamily-116/IOApplePartitionScheme.cpp.auto.html

Can be identified by header 0x455208 instead of the common 512 block size 0x455202

**Comment by bitsgalore** [comment]
OK, I think I figured this out, have a look at the changes in this branch:

<https://github.com/KBNLresearch/isolyzer/tree/issue-22>

This gives meaningful output for all of the images referred to in this thread. 

I've made a provisional 1.4 alpha release:

<https://github.com/KBNLresearch/isolyzer/releases/tag/1.4.0-alpha>

PyInstaller link:

<https://pypi.org/project/isolyzer/1.4.0a0/>

I want to do some more testing and clean-up before I turn this into a new stable release.

**Comment by thorsted** [comment]
This is looking great, does the trick on most of my sample files. Here is a test file which now identifies with the PM marker, but HFS+ is not identified at offset 33792.
[10MB-CD-DVD.dmg.zip](https://github.com/KBNLresearch/isolyzer/files/8454027/10MB-CD-DVD.dmg.zip)

**Comment by bitsgalore** [comment]
@thorsted Now that's an interesting file. According to the info in the zero block, the image has a 2048 byte block size, but at the same time the actual partition maps follow a 512 byte pattern (first PM at offset 512, 2nd at 1024, etc.). If I manually override the 2048 value in the code, and replace it by 512 I get the following output:

```xml
<?xml version="1.0" ?>
<isolyzer>
    <toolInfo>
        <toolName>cli.py</toolName>
        <toolVersion>1.4.0-alpha</toolVersion>
    </toolInfo>
    <image>
        <fileInfo>
            <fileName>10MB-CD-DVD.iso</fileName>
            <filePath>/home/johan/kb/iso-identification/10MB-CD-DVD.iso</filePath>
            <fileSizeInBytes>2532864</fileSizeInBytes>
            <fileLastModified>Tue Mar  8 14:24:20 2022</fileLastModified>
        </fileInfo>
        <statusInfo>
            <success>True</success>
        </statusInfo>
        <sectorOffset>0</sectorOffset>
        <tests>
            <containsKnownFileSystem>True</containsKnownFileSystem>
            <sizeExpected>2531328</sizeExpected>
            <sizeActual>2532864</sizeActual>
            <sizeDifference>1536</sizeDifference>
            <sizeDifferenceSectors>0.75</sizeDifferenceSectors>
            <sizeAsExpected>False</sizeAsExpected>
            <smallerThanExpected>False</smallerThanExpected>
        </tests>
        <fileSystems>
            <fileSystem TYPE="HFS">
                <appleZeroBlock>
                    <signature>ER</signature>
                    <blockSize>2048</blockSize>
                    <blockCount>1236</blockCount>
                    <deviceType>0</deviceType>
                    <deviceID>0</deviceID>
                    <driverData>0</driverData>
                    <driverDescriptorCount>0</driverDescriptorCount>
                    <driverDescriptorBlockStart>0</driverDescriptorBlockStart>
                    <driverDescriptorBlockCount>0</driverDescriptorBlockCount>
                    <driverDescriptorSystemType>0</driverDescriptorSystemType>
                </appleZeroBlock>
                <applePartitionMap>
                    <signature>PM</signature>
                    <numberOfPartitionEntries>2</numberOfPartitionEntries>
                    <partitionBlockStart>1</partitionBlockStart>
                    <partitionBlockCount>63</partitionBlockCount>
                    <partitionName>Apple</partitionName>
                    <partitionType>Apple_partition_map</partitionType>
                    <partitionLogicalBlockStart>0</partitionLogicalBlockStart>
                    <partitionLogicalBlockCount>63</partitionLogicalBlockCount>
                    <partitionFlags>3</partitionFlags>
                    <bootCodeBlockStart>0</bootCodeBlockStart>
                    <bootCodeSizeInBytes>0</bootCodeSizeInBytes>
                    <bootCodeLoadAddress>0</bootCodeLoadAddress>
                    <bootCodeJumpAddress>0</bootCodeJumpAddress>
                    <bootCodeChecksum>0</bootCodeChecksum>
                    <processorType/>
                </applePartitionMap>
                <applePartitionMap>
                    <signature>PM</signature>
                    <numberOfPartitionEntries>2</numberOfPartitionEntries>
                    <partitionBlockStart>64</partitionBlockStart>
                    <partitionBlockCount>4880</partitionBlockCount>
                    <partitionName>disk image</partitionName>
                    <partitionType>Apple_HFS</partitionType>
                    <partitionLogicalBlockStart>0</partitionLogicalBlockStart>
                    <partitionLogicalBlockCount>4880</partitionLogicalBlockCount>
                    <partitionFlags>1073741875</partitionFlags>
                    <bootCodeBlockStart>0</bootCodeBlockStart>
                    <bootCodeSizeInBytes>0</bootCodeSizeInBytes>
                    <bootCodeLoadAddress>0</bootCodeLoadAddress>
                    <bootCodeJumpAddress>0</bootCodeJumpAddress>
                    <bootCodeChecksum>0</bootCodeChecksum>
                    <processorType/>
                </applePartitionMap>
                <masterDirectoryBlock>
                    <signature>H+</signature>
                    <blockSize>9614</blockSize>
                    <blockCount>3729623060</blockCount>
                    <volumeName/>
                </masterDirectoryBlock>
            </fileSystem>
        </fileSystems>
    </image>
</isolyzer>
``` 
Could it be that the image is malformed (e.g. wrong block size in the zero block)?

**Comment by thorsted** [comment]
Mounts on my Mac. I created it using the Disk Utility tool. Disktype likes it. 
```
10MB-CD-DVD.dmg
Regular file, size 2.416 MiB (2532864 bytes)
Type code "rved", creator code "ksdd"
Apple partition map, 2 entries
Partition 1: 31.50 KiB (32256 bytes, 63 sectors from 1)
  Type "Apple_partition_map"
Partition 2: 2.383 MiB (2498560 bytes, 4880 sectors from 64)
  Type "Apple_HFS"
  HFS Plus file system
    Volume size 2.383 MiB (2498560 bytes, 610 blocks of 4 KiB)
    Volume name "10MB-CD-DVD"
```

**Comment by bitsgalore** [comment]
So I just had another look at the [Wikipedia entry on partition maps](https://en.wikipedia.org/wiki/Apple_Partition_Map), which led me to:

1. [Here](http://t2sde.org/handbook/html/t2.partition.apple.html), which says "While the original intent was to handle various block sizes, in practice only 512 byte blocks are supported."
2. [Thus technical note from Apple](https://developer.apple.com/library/archive/technotes/tn2166/_index.html), which says "Do not assume that the block size is always going to be 512 bytes."

If the pattern in this file is widespread, it means that the changes will introduce a regression for images for which isolyzer previously worked. To make it work for all cases, an option would be to:

1. Iterate over the partition maps assuming a block size of 512 bytes.
2. If the block size as defined in the zero block is some other value than 512, repeat the iteration using this value.
3. Define some criteria to decide on whether to report the result of either 1 or 2.

I'll have a further look into this.

**Comment by thorsted** [comment]
Interesting. I have never used pdisk before, but ran it while the 10MB-CD-DVD.dmg and Beatles disks were mounted.

```
Partition map (with 512 byte blocks) on '/dev/rdisk3'
 #:                type name       length   base   ( size )
 1: Apple_partition_map Apple          63 @ 1     
 2:           Apple_HFS disk image   4880 @ 64     (  2.4M)

Device block size=2048, Number of Blocks=1236 (2.4M)
DeviceType=0x0, DeviceId=0x0


Partition map (with 512 byte blocks) on '/dev/rdisk4'
 #:                type name      length   base    ( size )
 1: Apple_partition_map Apple          7 @ 1      
 2:           Apple_HFS MR_PAJAC 1331199 @ 8       (650.0M)

Device block size=512, Number of Blocks=0 (0.0 )
DeviceType=0x0, DeviceId=0x0
```
Both state a 512 byte block size, but the device block size is different.

**Comment by bitsgalore** [comment]
Also found this:

<https://www.forensicfocus.com/forums/general/apple-partition-map-512-bytes-vs-2048-bytes-e-g-on-cd/>

and this might cover something similar:

<https://forum.winworldpc.com/discussion/9772/mac-os-8-6-iso-is-corrupt>

**Comment by thorsted** [comment]
Here is another fun APM sample. Sorry to keep compounding the issue.
https://archive.org/details/appledevelop1989

```
Partition map (with 512 byte blocks) on '/dev/rdisk2'
 #:                type name      length   base   ( size )
 1: Apple_partition_map Apple         63 @ 1     
 2:        Apple_Driver*Macintosh     32 @ 64    
 3:           Apple_HFS MacOS      20000 @ 96     (  9.8M)
 4:          Apple_Free Extra     136274 @ 20096  ( 66.5M)

Device block size=512, Number of Blocks=156370 (76.4M)
DeviceType=0x1, DeviceId=0x1
Drivers-
1:  10 @ 64, type=0x1
```

**Comment by bitsgalore** [comment]
So in the end I applied this very simple fix which should do the trick (could be easily extended with more offsets if needed, but so far I've only seen 512 and 2048, so I think this should cover it):

<https://github.com/KBNLresearch/isolyzer/commit/96ac00e468e51afe3b1461cb54378a1a9f5a34a6>

**Comment by bitsgalore** [comment]
@thorsted I had a look at your Apple_Develop_1989.iso image. What's a bit odd is is that based on the info in the zero block I would expect a much larger file. It also made me aware of a small bug that mixes up the block size and block count for the master directory block, so I just [fixed that](https://github.com/KBNLresearch/isolyzer/commit/d6cb99c7a51c4a7625642dcb1434daad8b051409).

Other than that the Isolyzer output looks OK I think, but let me know if you think anything is still missing.

**Comment by bitsgalore** [comment]
All fixed in v1.4 now.
